### DISCUSSION TITLE
Auto-Healing process

### DISCUSSION THREAD

**Discussion by dstoyanoff** [directus/directus#8652] [Feature Requests]
Hey guys,
I was thinking about something today. Would be amazing of Directus can run some cron process to heal the database configuration (directus_* tables) to prevent inconsistent behavior. I am sharing 2 cases where I've experienced that. The end result is that if you snapshot a corrupted instance, then it would fail to restore, since the schema process is quite strict.

**Issue 1:**
Once upon a time, in a galaxy far away, when there was no shapshot, we would manage everything using migrations. Those migrations were error prone, resulting in having trash data - e.g. relation not deleted from the relations table or duplicated field definitions. Those are things that do not cause issues on the UI, but internally do cause inconsistency of the data.

**Issue 2:**
This one is more of a present issue. Imagine that I drop a table from the database and did not remember to remove this table from the system tables' definitions. That would leave some records hanging there. If I create a snapshot and later on try to restore it, it would fail, because that table does not exist.

**Comment by rijkvanzanten** [comment]
Love it! Would be really useful to have a `directus database doctor` command to check for / fix any corrupted rows. The most difficult question now becomes: What known cases do we catch for? One that comes to mind is directus-relations rows that don't have a matching foreign key constraint in the DB 🤔

**Reply by dstoyanoff** [reply]
I think the main use case would be for users that occasionally do changes in the database directly, which might break Directus. The app itself is not immune to causing issues as well, if there is some bug, so it would be useful too.

**Reply by rijkvanzanten** [reply]
Totally. The difficulty is that we need a set of rules to check against 🤔

**Reply by beasteers** [reply]
I definitely see how difficult this could be - but I do think it's quite important because I realized how easy it is to corrupt Directus the other day and it was a pain to get back from.

Perhaps something interactive that shows the user inconsistencies - e.g.
 - collections table doesn't match available tables
 - fields that don't exist in the schema
 - relations that don't have a corresponding table

Perhaps this could be visible from the data model pages? Like a collection without a corresponding table could be marked as red and have an option to either delete the collection or recreate it from the information stored in directus. And same with fields? Right now if a collection exists without a table it shows up as a folder and may or may not throw an error when you try to delete.

I think this also ties into that feature request about renaming field names (and migrating field types) because one of the issues I had was that I had renamed a column a while back and directus still had traces of the old name in the snapshot (the metadata was split between the two columns). Having to do things like this manually while having the directus configuration being so brittle is a bit of a minefield

Having a CLI version like you mentioned is a great idea <3

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
